libqaul/http-api/src/auth/error.rs

```diff
@@ -0,0 +1,137 @@
+use crate::{
+    JSONAPI_MIME,
+    models::ConversionError,
+};
+use libqaul::QaulError;
+use identity::ID_LEN;
+use iron::{
+    IronError,
+    status::Status,
+};
+use json_api::{
+    Document,
+    Error,
+    ErrorSource,
+    ObjectConversionError,
+};
+use std::{
+    error::Error as StdError,
+    fmt::{Display, Formatter, Result},
+};
+
+#[derive(Debug)]
+pub (crate) enum AuthError {
+    MultipleData,
+    NoData,
+    ConversionError(ObjectConversionError),
+    NoSecret,
+    InvalidIdentity(ConversionError),
+    QaulError(QaulError),
+    NotLoggedIn,
+}
+
+impl AuthError {
+    fn detail(&self) -> String {
+        match self {
+            AuthError::MultipleData => 
+                "Multiple data were provided when the endpoint expects exactly one".into(),
+            AuthError::NoData => "Document contains no data".into(),
+            AuthError::ConversionError(e) => format!("Error converting generic object ({})", e),
+            AuthError::NoSecret => "No secret provided".into(),
+            AuthError::InvalidIdentity(e) => format!("Conversion Error ({})", e), 
+            AuthError::QaulError(e) => format!("Qaul Error ({:?})", e),
+            AuthError::NotLoggedIn => "Not logged in".into(),
+        }
+    }
+
+    fn into_error(&self) -> (Error, Status) {
+        let status = match self {
+            AuthError::QaulError(QaulError::NotAuthorised) => Status::Unauthorized,
+            AuthError::QaulError(QaulError::UnknownUser) => Status::NotFound,
+            AuthError::QaulError(QaulError::CallbackTimeout) => Status::InternalServerError,
+            AuthError::NotLoggedIn => Status::Unauthorized,
+            _ => Status::BadRequest,
+        };
+
+        let title = match self {
+            AuthError::MultipleData => Some("Multiple Data".into()),
+            AuthError::NoData => Some("No Data".into()),
+            AuthError::ConversionError(_) => Some("Object Error".into()),
+            AuthError::NoSecret => Some("No Secret".into()),
+            AuthError::InvalidIdentity(_) => Some("Invalid identity".into()),
+            AuthError::QaulError(QaulError::NotAuthorised) => Some("Not Authorized".into()),
+            AuthError::QaulError(QaulError::UnknownUser) => Some("Unknown User".into()),
+            AuthError::QaulError(QaulError::InvalidQuery) => Some("Invalid Query".into()),
+            AuthError::QaulError(QaulError::InvalidPayload) => Some("Invalid Payload".into()),
+            AuthError::QaulError(QaulError::CallbackTimeout) => None,
+            AuthError::NotLoggedIn => Some("Not Logged In".into()),
+        };
+
+        let detail = match self {
+            AuthError::ConversionError(ObjectConversionError::ImproperType{ expected, got }) => 
+                Some(format!("Primary data should be of type {} but is of type {} instead", 
```

Comment:
Not something that's really relevant for this PR, but maybe we should keep an eye on is translation services. With this much detail in the error messages, it'd be kinda cool if we could pull in translated strings somehow and let service authors (or us) be able to use some of them to directly display in the presentation layer.

...maybe

Reply:
that'd definitely be a good idea and probably falls under the "configuration management" area of my proposal so I'll definately have a think about it